### ISSUE TITLE
[CMIP6] multiple statistics from one download

### ISSUE THREAD

**Issue by tdcwilliams** [bopen/c3s-eqc-toolbox-template#24]
Hi Mattia (@malmans2),

I would like to get daily time series of different quantities calculated from the sea ice concentration using one download if possible.
The statistics I want to get for each day are for each model, and for Arctic and Antarctic:
- `sea_ice_extent = area_of_grid_cells[region_mask] * sea_ice_mask[region_mask]` with `sea_ice_mask = 1 if sea_ice_concentration > 0.3 else 0` and eg `region_mask_arctic = lat > 40` or `region_mask_antarctic = lat < -40`
- `sea _ice_area = area_of_grid_cells[region_mask] * sea_ice_concentration[region_mask]`

Are there already or could there be some functions implemented that:
- calculate an array with the area of all the grid cells?
- give a list of all the CMIP6 models?

Also could the `transform_func` argument of `download_and_transform` also take a list or even a dict of different transform functions (eg `transform_func=[get_arctic_extent, get_arctic_area, get_antarctic_extent, get_antarctic_area]`) and output a list/dict/pandas.DataFrame so we don't have to download everytime we want to make a calculate a different statistic?

**Comment by malmans2** [comment]
Hi @tdcwilliams,

None of the diagnostics/functions that you are looking for have been implemented yet.

It is not possible to pass multiple diagnostics functions at the moment, but the downloading step is cached separately under the hood. If you re-run `dawnload_and_transform` changing the `transform_func` only, no data is actually downloaded but we use cached data. So you should be able to easily build the list yourself:
```python
out = []
for transform_func in transform_funcs:
    out.append(download.download_and_transfrom(..., transform_func=transform_func))
```

If it's of use, we just added a much more advance notebook: https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/clima_and_bias_pr_cmip6_regionalised.ipynb

**Comment by tdcwilliams** [comment]
Hi Mattia @malmans2,

I've made a bit of progress with this
[cmip6_sea_ice_diagnostics.ipynb.zip](https://github.com/bopen/c3s-eqc-toolbox-template/files/11026747/cmip6_sea_ice_diagnostics.ipynb.zip)

However I am not so used to `xarray` and was not sure what to return for `get_sea_ice_extent` and `get_sea_ice_area`.
I just want time series, but don't know how to do this in `xarray`. I would also like to delete the sea ice concentration array which I don't need after the extent/area calculations.

Regards,
Tim

**Comment by tdcwilliams** [comment]
Thanks @malmans2 - it did help. I am still unclear about `sie` in my `get_sea_ice_extent` function though. So far I just turn it into a numpy array - should I be using some kind of xarray data type instead?
```
def get_sea_ice_extent(ds, sic_name, model, **kwargs):
    grid_cell_areas = get_grid_cell_areas(ds, **kwargs)
    sie = np.array([np.sum(grid_cell_areas[sic > SIC_THRESH])
                   for sic in ds[sic_name][:]])
    return ds.expand_dims(model=[model])
```

**Comment by tdcwilliams** [comment]
Hi @malmans2 - I have just got something downloaded and am playing around to get to know xarray a bit.
Unfortunately, it is not a regular lon-lat grid as I thought so the get_grid_cell_areas function is not quite right.

**Comment by tdcwilliams** [comment]
Hi @malmans2,
Could you take a look at this notebook please?
[cmip6_sea_ice_diagnostics.ipynb.zip](https://github.com/bopen/c3s-eqc-toolbox-template/files/11155601/cmip6_sea_ice_diagnostics.ipynb.zip)

I have the downloading and transforming working, to give some time series of sea ice extent and area.
 they are datasets with dimensions model, region (Arctic or Antarctic), time
I would like to add some time series plots at the end, showing the mean and spread of the models.
If you could show me how to do this for the Arctic and one experiment that would be great.

Also the calculation of the grid areas is maybe a bit complicated and doesn't work for either ERA5 or one CMIP6 model (they don't provide the corners of the grid cells), so maybe it would just be simpler to just regrid onto a 100km equal-area grid or something. Is regridding very slow?

Ciao and happy easter,
Tim

**Comment by malmans2** [comment]
Ciao @tdcwilliams !

I should be able to take a look this week.


> Also the calculation of the grid areas is maybe a bit complicated and doesn't work for either ERA5 or one CMIP6 model (they don't provide the corners of the grid cells), so maybe it would just be simpler to just regrid onto a 100km equal-area grid or something. Is regridding very slow?

We actually already have a function that is optimised for interpolations. It's called `diagnostics.regrid` (see [here](https://github.com/bopen/c3s-eqc-automatic-quality-control/blob/b463e8fd145844f6ba53f81f65cc9f9875a4df90/c3s_eqc_automatic_quality_control/diagnostics.py#L66-L71)).

Sorry, it was originally meant to be a private method (and I might move it under utils in the future), so I forgot to add docstrings. But it's quite simple and we use https://xesmf.readthedocs.io/en/latest/ under the hood. Here is the docstring:

- `obj` is the xarray object you want to regrid
- `grid_out` is the output grid
- `method` is the `xesmf` interpolation you want to use (e.g., bilinear, conservative, ...)
- `**kwargs` is any other keyword argument for `xesmf`

For example, it's used in a couple of notebook in `wp4` to interpolate CMIP to ERA5 (see [here](https://github.com/bopen/c3s-eqc-toolbox-template/tree/main/notebooks/wp4)).

Interpolating shouldn't be too slow at all, and if you use our regrid function it's optimised (we compute the weights only once, and we cache them on disk through a netcdf file. That way, anytime you need to interpolate data on the same grid the weights are re-used). So it's very much up to you, if you think it's scientifically OK interpolating, you can probably do that.

**Comment by tdcwilliams** [comment]
Thanks @malmans2, I might use that regrid method instead of trying to code up areas for yet another type of grid. Great to hear it is optimised. The model SICs are pretty smooth so it should probably be fine to interpolate them.
Ciao!

**Comment by malmans2** [comment]
Hi @tdcwilliams,

I'm looking at you notebook, sorry about the delay.
Quick question: Why are you computing the cell areas rather than using the model output?
Can we use the variable `Grid-cell area for ocean variables` that is available on the CDS?
See: https://cds.climate.copernicus.eu/cdsapp#!/dataset/projections-cmip6?tab=form

**Comment by malmans2** [comment]
Hi @tdcwilliams,

I'm looking at you notebook, sorry about the delay.
Quick question: Why are you computing the cell areas rather than using the model output?
Can we use the variable `Grid-cell area for ocean variables` that is available on the CDS?
See: https://cds.climate.copernicus.eu/cdsapp#!/dataset/projections-cmip6?tab=form

**Comment by tdcwilliams** [comment]
Hi @malmans2 - I hadn't noticed we could output the grid-cell areas - that is certainly a much easier way!

**Comment by malmans2** [comment]
Hi @tdcwilliams,

I've added a template for you use case. You can find it [here](https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/cmip6_sea_ice_diagnostics.ipynb).
It's just a starting point, please make sure the analysis is done correctly. We will update the template as you make progress.


A couple of comments:
1. I've executed it on WP4 in the VM, so you should find everything already cached there.
2. Unfortunately we have to compute the cell areas for ERA5. I've added a diagnostic to do that, you can check the code [here](https://github.com/bopen/c3s-eqc-automatic-quality-control/blob/e9d6cf8282c84de61322ec42608fe6e56431516d/c3s_eqc_automatic_quality_control/diagnostics.py#L381-L432).
3. I think the cell area might be missing for some CMIP model. Let me know if that's the case for some of the models you need. If we are lucky and they have regular grids, we can use the same diagnostic used for ERA5. Otherwise, we will have to re-grid or implement something similar to what you did in your notebook.
4. When you start analysing large time periods, you might want to explore different chunking (e.g., 10 years).

The template notebook should produce these figures:
![image](https://user-images.githubusercontent.com/22245117/231520671-cec19002-1863-4d76-854b-df55eae2f75b.png)
![image](https://user-images.githubusercontent.com/22245117/231521620-6f8d776f-531d-4924-8482-32e076e7a8b4.png)

**Comment by malmans2** [comment]
Hi Tim,

> maybe the diagnostics.regrid option might be the best "one-size-fits-all" approach?

I'll implement this in the template. I just need to know which grid to use for the interpolation. Maybe ERA5? I think that's what CMCC does in a template for WP4.

> most seem to be lost connection to CDS for download - would having larger chunks help this?

Not sure, unfortunately these issues are very hard to debug as we don't maintain the VM or `cdsapi`. They are also hard to reproduce, often they are just showing up when the system is busy. I think CMCC has been working OK with 10y chunking (see this [notebook](https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/trend_pr_cmip6.ipynb))

> another crash (kernel just restarts) is with an HR dataset CMCC-CM2-HR4 (I guess it is a high-resolution one?) maybe from running out of memory.
I tried transform_chunks=True but got an error; after that I did my own manual chunking (download_and_transform followed by xr.merge) and I was able to process this dataset.

I'll look into this. But I don't understand why something changed with `transform_chunks=True`.  Are you referring to the template I uploaded yesterday? That argument is not specified, so it should be True by default.

> I wonder for such large datasets like CMIP6 maybe most users wouldn't try to download and process on the fly (for me it is taking hours if not days) - probably they would download first and then do their processing? That is, perhaps a download script plus a notebook might be a better solution?

It's not really processed on the fly.
This is what's happening with `transform_chunks=True`:
1. download and save a raw chunk to disk
2. open the raw chunk
3. transform the raw chunk
4. save the transformed chunk to disk

If you want to download all data first, then process, you can just run `download_and_transform` with `transform_func=None`. E.g.:
```python
for transform_func in (None, my_transform_func):
    ds = download.download_and_transform(collection_id, request, transform_func=transform_func, **kwargs)
```


If it's not clear, with `transform_chunk=False` this is the workflow:
1. download and save ALL raw chunks to disk
2. open and merge ALL raw chunks
3. transformed the merged dataset
4. save the transformed dataset to disk
I would only use `transform_chunk=False` for operations that perform reductions on the same dimension of the chunks (e.g., dimension is time and you need to compute the climatology).

**Comment by malmans2** [comment]
Hi @tdcwilliams,

I've explored a bit the issue you mentioned in `CMCC-CM2-HR4`. The problem is that `siconc` and `areacello` have inconsistent sizes (I believe this is due to NEMO's halo points).

I've added this code in the [template](https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/cmip6_sea_ice_diagnostics.ipynb):
```python
        # Remove extra-points
        isel_dict = {}
        for dim, size in areacello.sizes.items():
            match size - siconc.sizes[dim]:
                case 1:
                    isel_dict[dim] = slice(None, -1)
                case 2:
                    isel_dict[dim] = slice(1, -1)
        if isel_dict:
            areacello = areacello.isel(**isel_dict).drop(list(isel_dict))
```

`CMCC-CM2-HR4` is now successfully plotted:
![image](https://user-images.githubusercontent.com/22245117/231776119-ab3849d4-6a5f-44c9-b8ff-91ab291aa75c.png)
![image](https://user-images.githubusercontent.com/22245117/231776150-9601fa1a-2f09-470c-a6ef-a6d46dde1e81.png)

**Comment by malmans2** [comment]
Hi @tdcwilliams,

I made some progress.
1. I've generalised the function to compute areas, so it works OK with both 1D and 2D lat/lon (see [here](https://github.com/bopen/c3s-eqc-automatic-quality-control/blob/d8e7f6d81dc458ee01dbb998ffbf3eff33a041bc/c3s_eqc_automatic_quality_control/diagnostics.py#L383-L437)). The areas are now cached, so they are computed only once when lat/lon bounds are the same.
2. I've changed the [notebook template](https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/cmip6_sea_ice_diagnostics.ipynb). Now the areas are always computed and never downloaded. Looks like it's working OK, except for a model `fio_esm_2_0` (I believe the issue is the order of the vertices provided, but I need to investigate)
3. I've tried a couple of years for all models, and there are a couple of models that are quite different (next week I'll make sure that there's no bug in the code that computes the areas).

![image](https://user-images.githubusercontent.com/22245117/232092611-b8df1f4f-08b3-4ded-8459-8b6360c3c125.png)
![image](https://user-images.githubusercontent.com/22245117/232092646-fbe8c3e2-623a-4f7e-abbf-cdda464637f8.png)


Let's catch up next week. You now have a few options:
1. Regrid all models to the same grid, then compute areas
2. Download areas when available, compute otherwise
3. Always compute areas on the native grid

Have a good weekend!

**Comment by tdcwilliams** [comment]
Ciao @malmans2 - thanks a lot - that looks great!
Sorry to be slow in getting back to you - I was a bit sick last week.

I think I am leaning towards (1) the regridding approach (simplicity of being able to treat all models the same and not needing many different cases).

A convenient target grid would be the ones (Arctic or Antarctic) for https://cds.climate.copernicus.eu/cdsapp#!/dataset/satellite-sea-ice-concentration?tab=overview (25-km equal-area grids, so each grid cell is just 625km^2). This would let also let us reuse the approach to add comparisons to the satellite observations later on.

Ciao, Tim

**Comment by malmans2** [comment]
Hi Tim,
I'll update the template using (1).
What kind of interpolation do you think is the best for sea ice concentration variables?
These are the interpolations available: https://xesmf.readthedocs.io/en/latest/notebooks/Compare_algorithms.html

**Comment by tdcwilliams** [comment]
Thanks Mattia. Probably `conservative` is the best interpolation method.

**Comment by malmans2** [comment]
Hi @tdcwilliams,

The template notebook is ready: https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/cmip6_sea_ice_diagnostics.ipynb

I think I had the units wrong in my previous figures, should be OK now.
I only tested a couple of models, I'm now running the same notebook for all CMIP models, but CDS queue is quite long and the moment.

![image](https://user-images.githubusercontent.com/22245117/232533586-656ade5d-c1af-4644-a4ac-9c84b0b350f4.png)
![image](https://user-images.githubusercontent.com/22245117/232533649-3c92cd54-53c9-4119-9690-b7d138eb07fd.png)

**Comment by malmans2** [comment]
Hi there,
I've updated the [template notebook](https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/cmip6_sea_ice_diagnostics.ipynb). It now works with all CMIP6 models and it's already cached on the VM, user WP4.

Note that the `conservative` interpolation also needs `ignore_degenerate=True` for several models.
The reason why ignore_degenerate is needed is explained here: https://github.com/JiaweiZhuang/xESMF/issues/60.
Not sure if this is not ideal and therefore you should use a `bilinear` interpolation instead (you can just change the `method` argument).

Here is a quick-and-dirty plot of the results:
![image](https://user-images.githubusercontent.com/22245117/232753582-94e14684-8b1c-4cee-8070-e2c491e99056.png)
![image](https://user-images.githubusercontent.com/22245117/232753663-0828b138-389b-4aa0-83fe-712e52e607fa.png)

**Comment by tdcwilliams** [comment]
Thanks @malmans2 - I'll try it out soon. Will try changing the method to `bilinear`; `nearest_s2d` could be another option, but maybe it doesn't matter too much.
Ciao, Tim

**Comment by malmans2** [comment]
Hi @tdcwilliams,

What's the status of this issue? Are you still working on this and/or are there any standing issues?

**Comment by tdcwilliams** [comment]
Hi @malmans2,
I am getting back to the CMIP6 work now.
I will try out the latest template and get back to you

**Comment by tdcwilliams** [comment]
Hi @malmans2,
I changed the plotting a bit to just show the range, tertiles and median of the CMIP6 models, as well as adding ERA5
[cmip6_sea_ice_diagnostics.ipynb.zip](https://github.com/bopen/c3s-eqc-toolbox-template/files/12546782/cmip6_sea_ice_diagnostics.ipynb.zip)

I think it would also be good to add (on the same plots) the extent and area from the 2 products in the satellite sea ice concentration dataset:
- one you used already for the regridding (`origin='esa_cci'`)
- the other is `origin='eumetsat_osi_saf'`

These are daily datasets - if the monthly mean was taken it would enable a direct comparison to be made. The monthly std could also be taken to give an idea of the variability in the satellite data.

There should then also be another 2x2 plots with errors called the IIEE (Integrated Ice Edge Error) and the Extent Bias.
These could be computed as follows:
```
sic_model = ... # SIC from model
sic_sat = ... # SIC from satellite
cell area = 25 ** 2 # km^2
model_ice = sic_model > sic_threshold
sat_ice = sic_sat > sic_threshold # monthly mean
good_pixels = np.isfinite(sic_model) * np.isfinite(sic_sat) # get common area that isn't masked (eg not land)
over = good_pixels * model_ice * (~sat_ice) # model overestimation
under = good_pixels * (~model_ice) * sat_ice # model underestimation
iiee = (over.sum() + under.sum()) * cell_area
extent_bias = (over.sum() - under.sum()) * cell_area 
```

We should also plot the ensemble statistics of the projected extent and area.
It could be on the same graph or maybe in a separate notebook (perhaps this is best because we also need to do some more work with the projections to show when different Arctic shipping routes would be ice-free)

Thanks a lot for your help and sorry for letting this go on for so long - I had to work on some other projects for a while.

**Comment by malmans2** [comment]
Got it.
If there's no rush I'll take a look in the next couple of weeks (I'm busy with the General Assembly next week)

**Comment by tdcwilliams** [comment]
Thanks, the aim is to finish by the end of the month.
I can see how far I can get next week by myself then. Maybe I can make a draft that you can run for the full period.

**Comment by malmans2** [comment]
Sure!
I might be able to take a look at the notebook before the GA, either this afternoon or Monday morning.

**Comment by tdcwilliams** [comment]
thanks - I was trying to download satellite SIC and got the errors
```
Exception: Client has not agreed to the required terms and conditions.. To access this resource, you first need to accept the termsof 'EUMETSAT OSI SAF sea ice concentration licence' at https://cds.climate.copernicus.eu/cdsapp/#!/terms/eumetsat-osi-saf-sic. To access this resource, you first need to accept the termsof 'ESA-CCI sea ice concentration product licence' at https://cds.climate.copernicus.eu/cdsapp/#!/terms/ESA-CCI-sea-ice-concentration
```
The command was:
```
request_sat_esa = ('satellite-sea-ice-concentration', {'origin': 'esa_cci', 'version': 'v2', 'cdr_type': 'cdr', 'variable': 'all', 'year': ['2010'], 'month': ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12'], 'day': ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31'], 'region': 'northern_hemisphere'})
ds = download.download_and_transform(
        *request_sat_esa,
        chunks = {"year": 1},
    )
```

**Comment by malmans2** [comment]
I don't know who's the owner of the default cdsapirc used for WP4, but it looks like they didn't agree to the terms and conditions.
You should use you own account as explained here: https://github.com/bopen/c3s-eqc-toolbox-template/issues/94

**Comment by tdcwilliams** [comment]
thanks, it works now.

**Comment by tdcwilliams** [comment]
Hi @malmans2, 
I'm making a bit of progress, so you can just concentrate on the GA and I'll send the latest notebooks next week.
Have a good weekend!

**Comment by malmans2** [comment]
I made some progress as well, mainly generalising the notebook to make it easier to implement the new features.
I'll send you something Monday morning before I leave for Brno.

I noticed that the satellite data are missing various years.
Which time range would you like to cover for this analysis? Do you want to just skip months with missing satellite data?

**Comment by tdcwilliams** [comment]
thanks @malmans2,
we should cover as much as possible. For diagnostic purposes the full
length of the experiments, for comparisons with satellite all the months
where sat data is available. I guess we can just skip the months with
missing data.
Have a good weekend,
Tim

**Comment by malmans2** [comment]
Hi @tdcwilliams,

I've revised a bit the notebook. I haven't implemented the new features yet, but it should be easy now to implement new diagnostics and add more data. I've also revised a little the plots.

I might be able to implement the new features during the GA. I'll let you know I make progress.

In the meantime, just a comment about satellite data. Unfortunately, the current version of cdsapi does not allow to explore the dataset to find out missing dates. We could use try/except, but that's a little risky. We might have to hardcode the available dates for satellite data.

Here is the latest template: https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/cmip6_sea_ice_diagnostics.ipynb
Here is the notebook executed: https://gist.github.com/malmans2/07518e17ad464a49e26cceeedf47298d

**Comment by tdcwilliams** [comment]
Thanks for the work @malmans2 - I'll take a look tomorrow, and add the missing dates for the satellite data.
Ciao, Tim

**Comment by tdcwilliams** [comment]
Hi @malmans2  - does it only crash if the whole month is missing? If so, they are
```
missing_months_esa = [
	'2011-11,'
	'2011-12,'
	'2012-01,'
	'2012-02,'
	'2012-03,'
	'2012-04,'
	'2012-05,'
	'2012-06,'
]

missing_months_eumetsat = [
	'1986-04,'
	'1986-05,'
]
```
I can give individual days if needed. By the way for ESA `2011-10` and `2012-07` could be added since they only have a few days each

**Comment by malmans2** [comment]
Good catch! If we use monthly chunks, looks like it crashes only if the whole month is empty (I think the other day I only tested 1986).

**Comment by tdcwilliams** [comment]
Hi @malmans2, I have added those missing dates (but not used them yet) and have added the observations to the plot (just for 2010 though). I am going to try to make an error calculating function now (if you haven't started it already).
Ciao, Tim
[cmip6_sea_ice_diagnostics.ipynb.zip](https://github.com/bopen/c3s-eqc-toolbox-template/files/12595028/cmip6_sea_ice_diagnostics.ipynb.zip)

**Comment by malmans2** [comment]
Sounds good! 

I haven't make any progress. Unfortunately, the WiFi is not working well here in Brno. I'll be back in the office on Friday.

**Comment by tdcwilliams** [comment]
Hi @malmans2, I've made a decent amount of progress but it's not quite clean enough to share. To save us duplicating work, let me tidy it up tomorrow and I'll have something that you could take a look at on Monday. Have a good Friday and weekend, Tim

**Comment by tdcwilliams** [comment]
Hi @malmans2 ,
here is the latest diagnostics notebook.
[cmip6_sea_ice_diagnostics.ipynb.zip](https://github.com/bopen/c3s-eqc-toolbox-template/files/12657356/cmip6_sea_ice_diagnostics.ipynb.zip)

Could you run it for the full period?

Before you do though, the ERA5 interpolation needs a fix for the longitude=180 line - see the attached figure

It was also getting some crazy values near the north pole (eg 360% concentration)
![era5_sic](https://github.com/bopen/c3s-eqc-toolbox-template/assets/2579521/c2628b2d-60bc-4045-a29d-b45e25d0d968)

We should add some more projection experiments too
- the available models are different for each experiment so I'll find out which ones are available for those ones (the other experiments only have very few models with SIC available)

About the evaluation - the notebooks will be clearer if we separate the evaluation. I'll make separate issues for those ones (sea ice concentration and thickness)

**Comment by tdcwilliams** [comment]
PS for the plotting - with the full range the seasonal cycle might be too squashed to be visible, so maybe we could have the current plots for present time +/- 10 or 20 years. For the full period, we could plot the yearly maximum and minimum of each of the diagnostics.
Thanks a lot for the help.

**Comment by malmans2** [comment]
OK! I should be able to take a look this afternoon.

For the interpolation issue, have you checked the other models as well?
Is the issue only happening with ERA5?

**Comment by tdcwilliams** [comment]
Hi @malmans2,
`access_cm2` also has the interpolation problem (max=165%) also
![sic_access_cm2](https://github.com/bopen/c3s-eqc-toolbox-template/assets/2579521/d02492ba-0d5d-4c85-a4dd-ea8f3f8e8e4f)

**Comment by malmans2** [comment]
Hi @tdcwilliams,

Thanks, that's helpful! I'm a little busy today, so I didn't look into this yet.
But I have some ideas. I suspect this is the issue we are facing: https://github.com/pangeo-data/xESMF/issues/157

I'll give it a go tomorrow.

**Comment by malmans2** [comment]
Hi @tdcwilliams,

I'll give it another go tomorrow, but looks like I can only fix the interpolation issue using the `periodic=True` argument of `xesmf`. The problem is that `periodic` is not used for conservative interpolations.

Would it be OK to use another interpolation, such as `bilinear` or `nearest`? If yes, which one would you like to use?

**Comment by tdcwilliams** [comment]
Hi @malmans2 - thanks very much, bilinear should be ok I think

**Comment by tdcwilliams** [comment]
Hi @malmans2,

Here is the latest notebook - I've tidied things up a bit and added the satellite SIC as the code is not too intrusive.

[cmip6_sat_sea_ice_diagnostics.ipynb.zip](https://github.com/bopen/c3s-eqc-toolbox-template/files/12698486/cmip6_sat_sea_ice_diagnostics.ipynb.zip)

When you're happy with the interpolation, it would be great if you could cache the full time series.

Thanks a lot,
Tim

**Comment by malmans2** [comment]
Thanks! I'm on it today, hopefully you'll be able to check some results on Monday.

Is the latest notebook including the diagnostics in #102? I.e., can we close #102?

**Comment by tdcwilliams** [comment]
that notebook in #102 is a bit out of date. I'll upload the latest version of the SIC eval to #102 which has some extra calculations of error statistics. Some of the code is duplicated unfortunately, but it might still be good to have separate notebooks to stop this one being too long.

**Comment by malmans2** [comment]
OK, makes sense!

**Comment by malmans2** [comment]
Hi Tim,

I'm scraping the CDS forms because most of the experiments have some inconsistency?
Which experiments would you like to analyse? 
All of these, more, or less? `{"historical", "ssp1_2_6", "ssp2_4_5", "ssp3_7_0", "ssp5_8_5"}`

**Comment by malmans2** [comment]
Here is ERA5 with interpolation `method=bilinear` and `periodic=True`
![image](https://github.com/bopen/c3s-eqc-toolbox-template/assets/22245117/771f6123-f5f6-4e4a-8d42-d168343a56d5)

**Comment by tdcwilliams** [comment]
Hi @malmans2 - I added another issue for evaluation of sea ice thickness: #103.
I don't think we can manage to get these notebooks finished before Monday so I've asked for an extension.

**Comment by malmans2** [comment]
Hi @tdcwilliams,

OK. The extension is good, although we might be able to get it done before the deadline.

> Hi Tim,
> 
> I'm scraping the CDS forms because most of the experiments have some inconsistency. Which experiments would you like to analyse? All of these, more, or less? `{"historical", "ssp1_2_6", "ssp2_4_5", "ssp3_7_0", "ssp5_8_5"}`

Are these the experiments that you need to analysise?

**Comment by tdcwilliams** [comment]
Hi @malmans2 - that's right - those are the experiments with a significant number of models having sea ice concentration

**Comment by tdcwilliams** [comment]
Hi @malmans2 - the new deadline for the task is 30 November

**Comment by malmans2** [comment]
OK. BTW, it's now caching the last experiment.

**Comment by tdcwilliams** [comment]
great!

**Comment by malmans2** [comment]
Hi @tdcwilliams,

The notebook is ready.
Here is the template: https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/cmip6_sea_ice_diagnostics.ipynb
Here is the notebook executed: https://gist.github.com/malmans2/07518e17ad464a49e26cceeedf47298d

I resampled the timeseries (yearly means), otherwise the seasonal variability makes the plot too messy.
ERA5 in Antarctic looks odd, let me know if this is unexpected and there's something wrong in the template.

**Comment by malmans2** [comment]
> PS for the plotting - with the full range the seasonal cycle might be too squashed to be visible, so maybe we could have the current plots for present time +/- 10 or 20 years. For the full period, we could plot the yearly maximum and minimum of each of the diagnostics. Thanks a lot for the help.

Ooops, I forgot about this. Implementing it now.

**Comment by tdcwilliams** [comment]
Thanks a lot @malmans2, I'll play around a bit with the plotting.
Ciao, Tim

**Comment by tdcwilliams** [comment]
PS ERA5 does look weird in Antarctica. It looks a bit like there are missing years in the time series that just got joined up with straight lines.

**Comment by malmans2** [comment]
I'll run some check! (probably tomorrow)

**Comment by tdcwilliams** [comment]
Hi @malmans2, I have looked at it more closely myself and there is data in those flat periods, but maybe they have done some artificial constraining of the sea ice (eg with a climatology) when there was no data to assimilate (pre-1979) which could be why there is so little variability at times. I can't find anything about it in the documentation though. Perhaps we just stick to after 1980 in our plots.

Here's my latest notebook.
I've changed the plotting a bit to make things a bit clearer eg just plot interquartile range instead of tertiles and full range (on the advice of Lorenzo).

I kind of prefer a couple of extra functions `get_sic` and `get_output_grid` to make the cache function clearer so I added them.

I've also added extent and area for satellites.
However the missing data messes up the yearly max and min - do you know how to skip years with missing data?
(I guess somehow getting `xarray` to use `max` instead of `nanmax` would do it?)
[cmip6_sea_ice_diagnostics.ipynb.zip](https://github.com/bopen/c3s-eqc-toolbox-template/files/12747902/cmip6_sea_ice_diagnostics.ipynb.zip)

**Comment by malmans2** [comment]
Hi @tdcwilliams,

> Perhaps we just stick to after 1980 in our plots.

OK

> I've changed the plotting a bit to make things a bit clearer eg just plot interquartile range instead of tertiles and full range (on the advice of Lorenzo).

OK, I'll implement the same in the template

> I kind of prefer a couple of extra functions get_sic and get_output_grid to make the cache function clearer so I added them.

Fine by me. But to do so, we need to invalidate the cache and re-run the transform functions. If you are OK with it, I'll go ahead (but please don't run this notebook until everything is cached again)

BTW, are you sure that all models have the CF attribute `"standard_name"`? If yes, we can replace you code to get sic with this: `da_sic = ds.cf["sea_ice_area_fraction"]`

> I've also added extent and area for satellites.
However the missing data messes up the yearly max and min - do you know how to skip years with missing data?
(I guess somehow getting xarray to use max instead of nanmax would do it?)

OK, I'll add satellite data in the template

**Comment by tdcwilliams** [comment]
Hi @malmans2,
OK, it seems like `get_sic` works when I used it on models and satellite data, but your way is quicker if it also works.
If you are rerunning can you change the units to "$10^6$km$^2$" (so km is not in italics)?

**Comment by tdcwilliams** [comment]
Hi @malmans2, is it OK for me to run this notebook again?
Ciao, Tim

**Comment by malmans2** [comment]
Hi @tdcwilliams,

Not yet, I'm caching everything right now. Hopefully it will be ready in the afternoon.
Could you please check if the way the functions are coded is more clear?
https://github.com/bopen/c3s-eqc-toolbox-template/blob/main/notebooks/wp4/cmip6_sea_ice_diagnostics.ipynb

**Comment by tdcwilliams** [comment]
Thanks @malmans2. I think the code is very nice now. The satellite code is certainly much simpler with yearly chunking - is that working OK? Also, how does filling missing months with 0 avoid the max/min problem for the observations?

**Comment by malmans2** [comment]
> The satellite code is certainly much simpler with yearly chunking - is that working OK? 

It works OK, besides for EUMETSAT April/May 1986 (no-data which results in 0s). This is why I'm getting rid of those months: `ds = ds.where(ds != 0).dropna("time")`

> Also, how does filling missing months with 0 avoid the max/min problem for the observations?

That's a separate issue, which is addressed in this function:

```python
def full_year_only_resample(ds, reduction):
    mask = ds["time"].resample(time="Y").count() == 12
    return getattr(ds.resample(time="Y"), reduction)().where(mask, drop=True)
```

**Comment by tdcwilliams** [comment]
great, sounds good

**Comment by malmans2** [comment]
Hi @tdcwilliams,

I'm planning to work on #102 now.
Let me know if this notebook is OK as we'll probably re-use most of the code.

**Comment by tdcwilliams** [comment]
Hi @malmans2, it seems good.
Cheers,
Tim
